pkg/karmadactl/util/apiclient/apiclient_test.go

```diff
@@ -0,0 +1,201 @@
+				return verifyRestConfig(restConfig, caCert)
+			},
+			cleanup: func(kubeConfigPath string) error {
+				if err := os.Remove(kubeConfigPath); err != nil {
+					return fmt.Errorf("failed to clean up config file %s, got error: %v", kubeConfigPath, err)
+				}
+				return nil
+			},
+			wantErr: false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			caCert, err := test.prep(test.kubeConfigPath)
+			if err != nil {
+				t.Fatalf("failed to prep test environment, got error: %v", err)
+			}
+			defer func() {
+				if err := test.cleanup(test.kubeConfigPath); err != nil {
+					t.Errorf("failed to cleanup test environment, got error: %v", err)
+				}
+			}()
+			restConfig, err := RestConfig(test.context, test.kubeConfigPath)
+			if err == nil && test.wantErr {
+				t.Fatal("expected an error, but got none")
+			}
+			if err != nil && !test.wantErr {
+				t.Errorf("unexpected error, got: %v", err)
+			}
+			if err != nil && test.wantErr && !strings.Contains(err.Error(), test.errMsg) {
+				t.Errorf("expected error message %s to be in %s", test.errMsg, err.Error())
+			}
+			if err := test.verify(restConfig, caCert); err != nil {
+				t.Errorf("failed to verify creating rest config, got error: %v", err)
+			}
+		})
+	}
+}
+
+func prepRestConfig(kubeConfig, kubeConfigPath string) (string, error) {
+	caCert, privateKey, err := generateTestCACertificate()
+	if err != nil {
+		return "", fmt.Errorf("failed to generate CA certificate: %v", err)
+	}
+	base64CACert := base64.StdEncoding.EncodeToString([]byte(caCert))
+	base64PrivateKey := base64.StdEncoding.EncodeToString([]byte(privateKey))
+	kubeconfigFormatted := fmt.Sprintf(kubeConfig, base64CACert, base64CACert, base64PrivateKey)
+	if err := os.WriteFile(kubeConfigPath, []byte(kubeconfigFormatted), 0600); err != nil {
+		return "", fmt.Errorf("failed to write kubeconfig to file: %v", err)
+	}
+	return base64CACert, nil
+}
+
+func verifyRestConfig(restConfig *rest.Config, caCert string) error {
+	if restConfig == nil {
+		return errors.New("expected rest config, but got nil")
+	}
+	if restConfig.Host != "https://127.0.0.1:6443" {
+		return fmt.Errorf("expected rest config host to be %s, but got %s", "https://127.0.0.1:6443", restConfig.Host)
+	}
+
+	caCertDecoded, err := base64.StdEncoding.DecodeString(caCert)
+	if err != nil {
+		return fmt.Errorf("failed to decode ca certificate, got error: %v", err)
+	}
+	if !bytes.Equal(restConfig.TLSClientConfig.CAData, caCertDecoded) {
+		return fmt.Errorf("expected ca cert bytes %v to be %v", restConfig.TLSClientConfig.CAData, caCertDecoded)
+	}
+
+	return nil
+}
+
+// generateTestCACertificate generates a self-signed CA certificate and its associated RSA private key.
+// The certificate and private key are returned as PEM-encoded strings.
+func generateTestCACertificate() (string, string, error) {
+	// Generate a new RSA private key
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
```

Comment:
@mohamedawnallah In file [kubeconfig_test.go](https://github.com/karmada-io/karmada/blob/ff7dc4af44312bc7eafa05fbf1cd9723ac0cc2e7/operator/pkg/util/kubeconfig_test.go#L215-L252), I found functions with similar functionalities that are applicable across multiple test cases. For these reusable common functions, what best practices are recommended to optimize their organization and usage?
https://github.com/karmada-io/karmada/blob/ff7dc4af44312bc7eafa05fbf1cd9723ac0cc2e7/operator/pkg/util/kubeconfig_test.go#L215-L252

Reply:
Great finding! We could make the `GenerateTestCACertificate` function reusable by placing it in the [pkg/util/testing](https://github.com/karmada-io/karmada/tree/master/pkg/util/testing) package. I've submitted a patch to reuse the `GenerateTestCACertificate` function in both `pkg/karmadactl/util/apiclient_test.go` and `operator/pkg/util/kubeconfig_test.go`. Let me know if this approach works for you! 🙂

Reply:
I'm curious if placing it in pkg/util/testing might decrease coverage, as it could be recognized as part of the functional code.

Reply:
Hmm, I’m not sure about the impact on coverage since it’s just one function. I’d love to hear if you have any other location suggestions that might work better. 🙏

Reply:
I agree that extracting common functions is useful for testing, but do these functions also need test cases written for them?

Reply:
I don't think it's necessary to write test cases for the `GenerateTestCACertificate` function, as it will only be used in the test environment, especially since it will be triggered continuously in CI. However, I think it would be nice to have them. Perhaps we can consider adding those test cases in a follow-up PR later. What do you think?